x-pack/filebeat/tests/integration/input_reload_test.go

```diff
@@ -0,0 +1,354 @@
+				LogLevel:       proto.UnitLogLevel_DEBUG,
+				Config: &proto.UnitExpectedConfig{
+					Id:   "default",
+					Type: "elasticsearch",
+					Name: "elasticsearch",
+					Source: requireNewStruct(t,
+						map[string]interface{}{
+							"type":     "elasticsearch",
+							"hosts":    []interface{}{"http://localhost:9200"},
+							"username": "admin",
+							"password": "testing",
+							"protocol": "http",
+							"enabled":  true,
+						}),
+				},
+			},
+			{
+				Id:             "input-unit-1",
+				Type:           proto.UnitType_INPUT,
+				ConfigStateIdx: 1,
+				State:          proto.State_HEALTHY,
+				LogLevel:       proto.UnitLogLevel_DEBUG,
+				Config: &proto.UnitExpectedConfig{
+					Id:   "log-input",
+					Type: "log",
+					Name: "log",
+					Streams: []*proto.Stream{
+						{
+							Id: "log-input-1",
+							Source: requireNewStruct(t, map[string]interface{}{
+								"enabled": true,
+								"type":    "log",
+								"paths":   []interface{}{logFilePath},
+							}),
+						},
+					},
+				},
+			},
+		},
+		{
+			{
+				Id:             "output-unit",
+				Type:           proto.UnitType_OUTPUT,
+				ConfigStateIdx: 1,
+				State:          proto.State_HEALTHY,
+				LogLevel:       proto.UnitLogLevel_DEBUG,
+				Config: &proto.UnitExpectedConfig{
+					Id:   "default",
+					Type: "elasticsearch",
+					Name: "elasticsearch",
+					Source: requireNewStruct(t,
+						map[string]interface{}{
+							"type":     "elasticsearch",
+							"hosts":    []interface{}{"http://localhost:9200"},
+							"username": "admin",
+							"password": "testing",
+							"protocol": "http",
+							"enabled":  true,
+						}),
+				},
+			},
+			{
+				Id:             "input-unit-2",
+				Type:           proto.UnitType_INPUT,
+				ConfigStateIdx: 1,
+				State:          proto.State_HEALTHY,
+				LogLevel:       proto.UnitLogLevel_DEBUG,
+				Config: &proto.UnitExpectedConfig{
+					Id:   "log-input",
+					Type: "log",
+					Name: "log",
+					Streams: []*proto.Stream{
+						{
+							Id: "log-input-2",
+							Source: requireNewStruct(t, map[string]interface{}{
+								"enabled": true,
+								"type":    "log",
+								"paths":   []interface{}{logFilePath},
+							}),
+						},
+					},
+				},
+			},
+		},
+	}
+
+	// Once the desired state is reached (aka Filebeat finished applying
+	// the policy changes) we still wait for a little bit before sending
+	// another policy. This will allow the input to run and get some data
+	// into the publishing pipeline.
+	//
+	// nextState is a helper function that will keep cycling through both
+	// elements of the `units` slice. Once one is fully applied, we wait
+	// at least 10s then send the next one.
+	idx := 0
+	waiting := false
+	when := time.Now()
+	nextState := func() {
+		if waiting {
+			if time.Now().After(when) {
+				idx = (idx + 1) % len(units)
+				waiting = false
+				return
+			}
+			return
+		}
+		waiting = true
+		when = time.Now().Add(10 * time.Second)
```

Comment:
Why 10 seconds? Can this be faster? Will this test be flaky?

Reply:
Unfortunately not. On my machine the tests take 40s to run. It's not super long for an integration tests. Even if I reduce this constant to 500ms it still takes 40s (the first run is the current 10s wait, the second is a 500ms wait):
```
tiago@millennium-falcon beats/x-pack/filebeat/tests/integration  v1.19.6  fix-input-reload-under-agent [$!] % go test -tags=integration -v ./...
=== RUN   TestInputReloadUnderElasticAgent                                                                      
    input_reload_test.go:62: Temporary directory: /home/tiago/devel/beats/x-pack/filebeat/build/integration-tests/TestInputReloadUnderElasticAgent-1685462672
    input_reload_test.go:68: Temporary directory '/home/tiago/devel/beats/x-pack/filebeat/build/integration-tests/TestInputReloadUnderElasticAgent-1685462672' removed
--- PASS: TestInputReloadUnderElasticAgent (40.37s)
PASS
ok      github.com/elastic/beats/v7/x-pack/filebeat/tests/integration   40.565s
tiago@millennium-falcon beats/x-pack/filebeat/tests/integration  v1.19.6  fix-input-reload-under-agent [$] % go test -tags=integration -v ./...
=== RUN   TestInputReloadUnderElasticAgent
    input_reload_test.go:62: Temporary directory: /home/tiago/devel/beats/x-pack/filebeat/build/integration-tests/TestInputReloadUnderElasticAgent-1685462799
    input_reload_test.go:68: Temporary directory '/home/tiago/devel/beats/x-pack/filebeat/build/integration-tests/TestInputReloadUnderElasticAgent-1685462799' removed
--- PASS: TestInputReloadUnderElasticAgent (40.37s)
PASS
ok      github.com/elastic/beats/v7/x-pack/filebeat/tests/integration   40.588s
tiago@millennium-falcon beats/x-pack/filebeat/tests/integration  v1.19.6  fix-input-reload-under-agent [$!] % 
```